src/test/scala/edu/berkeley/ce/rockslicing/JointGeneratorSpec.scala

```diff
@@ -0,0 +1,347 @@
+
+    val expectedRockVolume = Seq[Face](
+      Face(Array(0.0, -1.0, 0.0), 1.0, 30.0, 0.0),
+      Face(Array(0.0, 1.0, 0.0), 1.0, 30.0, 0.0),
+      Face(Array(-1.0, 0.0, 0.0), 1.0, 30.0, 0.0),
+      Face(Array(1.0, 0.0, 0.0), 1.0, 30.0, 0.0),
+      Face(Array(0.0, 0.0, -1.0), 1.0, 30.0, 0.0),
+      Face(Array(0.0, 0.0, 1.0), 1.0, 30.0, 0.0)
+    )
+    assert((expectedOrigin sameElements generatedInput.origin) &&
+      (expectedLowerLeftCorner sameElements generatedInput.lowerLeftCorner) &&
+      (expectedUpperRightCorner sameElements generatedInput.upperRightCorner) &&
+      expectedRockVolume == generatedInput.rockVolume.map(_.roundToTolerance())
+    )
+  }
+
+  test("Master joint should be x-y plane centered at (-1.0, -1.0, -1.0)") {
+    val globalOrigin = Array[Double](0.5, 0.5, 0.5)
+    val boundingBox = (Array[Double](-1.0, -1.0, -1.0), Array[Double](3.0, 3.0, 3.0))
+    val rockVolume = Seq[InputFace](
+      InputFace(0.0, 90.0, Array(0.0, 0.0, 0.0), 30.0, 0.0),
+      InputFace(0.0, 90.0, Array(0.0, 2.0, 0.0), 30.0, 0.0),
+      InputFace(90.0, 90.0, Array(0.0, 0.0, 0.0), 30.0, 0.0),
+      InputFace(90.0, 90.0, Array(2.0, 0.0, 0.0), 30.0, 0.0),
+      InputFace(0.0, 0.0, Array(0.0, 0.0, 0.0), 30.0, 0.0),
+      InputFace(0.0, 0.0, Array(0.0, 0.0, 2.0), 30.0, 0.0)
+    )
+    val jointData = Seq[JointSet](
+      JointSet(0.0, 0.0, 0.5, 100.0, 30.0, 0.0)
+    )
+    val generatedInput = JointGenerator(globalOrigin, boundingBox, rockVolume, jointData)
+
+    val expectedJoints = Seq(
+      Joint(Array(0.0, 0.0, -1.0), Array(0.5, 0.5, 0.5), Array(-1.0, -1.0, -1.0), phi = 30.0,
+        cohesion = 0.0, shape = Vector.empty)
+    )
+    assert(expectedJoints == generatedInput.masterJoints)
+  }
+
+  test("Master joint should have normal (sqrt(3.0)/2.0, 0.0, 1.0/2.0) and be centered at (-1.0, -1.0, -1.0)") {
+    val globalOrigin = Array[Double](0.5, 0.5, 0.5)
+    val boundingBox = (Array[Double](-1.0, -1.0, -1.0), Array[Double](3.0, 3.0, 3.0))
+    val rockVolume = Seq[InputFace](
+      InputFace(0.0, 90.0, Array(0.0, 0.0, 0.0), 30.0, 0.0),
+      InputFace(0.0, 90.0, Array(0.0, 2.0, 0.0), 30.0, 0.0),
+      InputFace(90.0, 90.0, Array(0.0, 0.0, 0.0), 30.0, 0.0),
+      InputFace(90.0, 90.0, Array(2.0, 0.0, 0.0), 30.0, 0.0),
+      InputFace(0.0, 0.0, Array(0.0, 0.0, 0.0), 30.0, 0.0),
+      InputFace(0.0, 0.0, Array(0.0, 0.0, 2.0), 30.0, 0.0)
+    )
+    val jointData = Seq[JointSet](
+      JointSet(270.0, 30.0, 0.5, 100.0, 30.0, 0.0)
+    )
+    val generatedInput = JointGenerator(globalOrigin, boundingBox, rockVolume, jointData)
+
+    val expectedJoints = Seq(
+      Joint(Array(-1.0/2.0, 0.0, -sqrt(3.0)/2.0), Array(0.5, 0.5, 0.5), Array(-1.0, -1.0, -1.0), phi = 30.0,
+        cohesion = 0.0, shape = Vector.empty)
+    )
+
+    val jointComparison = expectedJoints.zip(generatedInput.masterJoints) map { case (joint1, joint2) =>
+        compareJoints(joint1, joint2)
+    }
+
+    assert(!jointComparison.contains(false))
```

Comment:
You can simplify this to
```scala
assert(expectedJoints.zip(generatedInput.masterJoints).forAll { case (joint1, joint2) => compareJoints(joint1, joint2) })
```

At least in my experience, a lot of times when you have an assertion or a return involving a literal `true` or `false`, the code is actually more complicated than it needs to be.

Also, we shouldn't use the name `compareJoints`, which implies checking two joints for the relative ordering under some scheme, when this is really an equality check.